### PR TITLE
[10.0] Extract values from recompute_json (in se_binding)

### PR DESCRIPTION
**Pull request by acsonefho**
To control values during recompute_json, extract method to get values to update

### CODE REVIEW

File: connector_search_engine/models/se_binding.py [hunk 1]
```diff
@@ -60,12 +60,28 @@ def recompute_json(self, force_export=False):
...
-                    record.write(vals)
+                values = record._get_values_recompute_json(
+                    mapper, force_export=force_export)
+                if values:
+                    record.write(values)
+
+    @api.multi
+    def _get_values_recompute_json(self, mapper, force_export=False):
+        """
+        Get values for current recordset during recompute_json
```
Comment (rvalyi) [inline connector_search_engine/models/se_binding.py:78]: s/for current recordset/for the current record/
(if there is only one this a record and not a record set)

Comment (acsonefho) [reply connector_search_engine/models/se_binding.py:78]: Hi @rvalyi , thank you for your comment :)
IMO for this case, my record is into a `recordset` of size 1

**Review by Cedric-Pigeon — APPROVED** [review]

**Review by lmignon — APPROVED** [review]

**Comment by lmignon** [comment]
@sebastienbeau Can you merge this one. I don't have the right to merge on this repository.

File: connector_search_engine/models/se_binding.py [hunk 2]
```diff
@@ -60,12 +60,28 @@ def recompute_json(self, force_export=False):
...
+        :param mapper: export mapper
+        :param force_export: bool
+        :return: dict
+        """
+        self.ensure_one()
+        data = mapper.map_record(self).values()
+        if self.data != data or force_export:
```
Comment (sebastienbeau) [inline connector_search_engine/models/se_binding.py:86]: hum I think we will have an issue here.
Indeed the comparison between the new data and the existing one is done in the method.
In the case that you inherit this method to update the data (ex : the default value for the SEO title)
As the data have being tested here you change in the data will not impact the status.

Moreover if we want to update the data of the mapping we can already do it by adding a custom mapping so I do not think that we need to add a hook what do you think ? @lmignon @Cedric-Pigeon @acsonefho

Comment (acsonefho) [reply connector_search_engine/models/se_binding.py:86]: Hi @sebastienbeau 
I update the function to keep your logic with the sync_state and the data.
For the case of seo_title, we want to update it in the data but also into Odoo (so the field value).
So here the 2 `if` will detect changes on `data` and the `values` will also contains the value of `seo_title` (to update it if necessary).

File: connector_search_engine/models/se_binding.py [hunk 3]
```diff
@@ -60,12 +60,34 @@ def recompute_json(self, force_export=False):
...
-                        vals['sync_state'] = 'to_update'
-                    record.write(vals)
+                        values.update({
+                            'sync_state': 'to_update',
+                        })
+                if values:
+                    record.write(values)
+
+    @api.multi
+    def _get_values_recompute_json(self, mapper, force_export=False):
```
Comment (simahawk) [inline connector_search_engine/models/se_binding.py:76]: I'd rename this by prefixing the purpose: `_recompute_json_get_values`

File: connector_search_engine/models/se_binding.py [hunk 4]
```diff
@@ -60,12 +60,34 @@ def recompute_json(self, force_export=False):
...
+        """
+        Get values for current recordset during recompute_json
+        :param mapper: export mapper
+        :param force_export: bool
+        :return: dict
+        """
+        self.ensure_one()
+        data = mapper.map_record(self).values()
```
Comment (simahawk) [inline connector_search_engine/models/se_binding.py:84]: I think that this method should stay atomic and just do:
```py
return mapper.map_record(self).values()
```
Its role should be to return updated json data no matter what. 
Then in the calling method you decide what do w/ these values.

File: connector_search_engine/models/se_binding.py [hunk 5]
```diff
@@ -60,12 +60,34 @@ def recompute_json(self, force_export=False):
             mapper = work.component(usage='se.export.mapper')
             lang = work.index.lang_id.code
             for record in work.records.with_context(lang=lang):
-                data = mapper.map_record(record).values()
+                values = record._get_values_recompute_json(
+                    mapper, force_export=force_export)
+                data = values.get('data')
+                # Better to understand with 2 IF
                 if record.data != data or force_export:
```
Comment (simahawk) [inline connector_search_engine/models/se_binding.py:67]: TBH I don't think that this comparison is reliable... am I wrong or here you can have a dict w/ nested lists and dictionaries? Even if records' real values do not change they could be different in any case.
We should put a big TODO to remind us that we need a better comparison, maybe using an hash (AFAIR Laurent and/or Seb mentioned this at the sprint).

**Review by simahawk — CHANGES_REQUESTED** [review]

File: connector_search_engine/models/se_binding.py (hunk 5 above)
Comment (lmignon) [reply connector_search_engine/models/se_binding.py:67]: @simahawk For sure the comparison is not right. We should use a hash comparison instead.

File: connector_search_engine/models/se_binding.py (hunk 4 above)
Comment (lmignon) [reply connector_search_engine/models/se_binding.py:84]: :+1:

**Comment by sebastienbeau** [comment]
As seen during the code sprint the seo_title should be implemented as computed field and set in the exported (I may work on it this week to close all shopinvader PR).

So we do not need to hack this code anymore.
Moreover the code in the inherit was complex see the case here : 
https://github.com/akretion/odoo-shopinvader/pull/242/files#diff-70c22be7eed40745a53e5ae8581c4417R32

I think it simpler to keep the code like it is. Moreover we can still extend the mapper if really needed, so we do not need extract hook.
So for me the existing code that just calling the mapper and compare the result is enough
I am going to add a comment regarding the comparison commit of @simahawk 
If I miss something will reopen the PR

**Pull request closed without merging**